autopush/web/registration.py

```diff
@@ -0,0 +1,311 @@
+                                 status_code=400,
+                                 errno=108)
+        data['chid'] = data.get('chid')
+        if data.get('uaid'):
+            request_pref_header = {'www-authenticate': PREF_SCHEME}
+            try:
+                settings.router.get_uaid(data['uaid'])
+            except ItemNotFound:
+                raise InvalidRequest("UAID not found",
+                                     status_code=410,
+                                     errno=103)
+            if not data.get('auth'):
+                raise InvalidRequest("Unauthorized",
+                                     status_code=401,
+                                     errno=109,
+                                     headers=request_pref_header)
+            settings = self.context['settings']
+            try:
+                auth_type, auth_token = re.sub(
+                    r' +', ' ', data['auth'].strip()).split(" ", 2)
+            except ValueError:
+                raise InvalidRequest("Invalid Authentication",
+                                     status_code=401,
+                                     errno=109,
+                                     headers=request_pref_header)
+            if auth_type.lower() not in AUTH_SCHEMES:
+                raise InvalidRequest("Invalid Authentication",
+                                     status_code=401,
+                                     errno=109,
+                                     headers=request_pref_header)
+            if settings.bear_hash_key:
+                is_valid = False
+                for key in settings.bear_hash_key:
+                    test_token = generate_hash(key, data['uaid'])
+                    is_valid |= constant_time.bytes_eq(bytes(test_token),
+                                                       bytes(auth_token))
+                if not is_valid:
+                    raise InvalidRequest("Invalid Authentication",
+                                         status_code=401,
+                                         errno=109,
+                                         headers=request_pref_header)
+
+
+class RegistrationHandler(BaseWebHandler):
+    """Handle the Bridge services endpoints"""
+    cors_methods = "POST,PUT,DELETE"
+
+    #############################################################
+    #                    Cyclone HTTP Methods
+    #############################################################
+    @threaded_validate(RegistrationSchema)
+    def post(self, *args):
+        """HTTP POST
+
+        Endpoint generation and optionally router type/data registration.
+
+
+        """
+        self.start_time = time.time()
+        self.add_header("Content-Type", "application/json")
+        params = self.valid_input['params']
+        # If the client didn't provide a CHID, make one up.
+        # Note, valid_input may explicitly set "chid" to None
+        params["channelID"] = (self.valid_input.get("chid") or
```

Comment:
Because it may not be. The validator (which is called for all REST methods, if I understand correctly) pulls the CHID and verifies if it exists, but it may not since some REST methods note it as optional. POST provides a CHID if it's not present.